### PR TITLE
sync skillmap progression

### PR DESCRIPTION
**Pull request by eanders-MS**
Sync skillmap progress to cloud. It does this by hooking into the location where the skillmap loads and saves user state and injecting some state from the cloud. I *think* this is a good approach. Would appreciate feedback!

### CODE REVIEW

File: pxtlib/auth.ts [hunk 1]
```diff
@@ -20,7 +20,7 @@ namespace pxt.auth {
     const DEV_BACKEND_LOCALHOST_5500 = "http://localhost:5500"; // if running in Docker container
     const DEV_BACKEND_LOCALHOST_8080 = "http://localhost:8080"; // if not running in Docker
 
-    const DEV_BACKEND = DEV_BACKEND_PROD;
```
Comment (livcheerful) [inline pxtlib/auth.ts:23]: is this a change we want to commit, or just for testing?

File: pxtlib/util.ts [hunk 2]
```diff
@@ -333,6 +333,115 @@ namespace ts.pxtc.Util {
...
+                } else if (!Object.is(atObjA, atObjB)) {
+                    // Value exists in both, and they're not the same instance.
+                    if (typeof atObjA !== typeof atObjB) {
+                        // Type changed. Replace the whole subtree.
+                        diff.replace.push({
+                            op: 'replace',
+                            path: basePath.concat(key),
+                            value: atObjB
+                        });
+                    } else {
```
Comment (riknoll) [inline pxtlib/util.ts:390]: Nitpick:  should the clause here just be combined with the above case?

File: pxtlib/util.ts [hunk 3]
```diff
@@ -333,6 +333,115 @@ namespace ts.pxtc.Util {
...
+            const lastKey: any = path.pop();
+            if (lastKey == null) {
+                throw new Error("jsonPath: missing last key");
+            }
+            let subObj = obj;
+            // Add sub-objects for each path element except the last one.
+            let currKey: any;
+            while (((currKey = path.shift())) != null) {
+                if (!(currKey in subObj)) {
+                    subObj[currKey] = {};
```
Comment (riknoll) [inline pxtlib/util.ts:432]: will this work with arrays? should recreate all of the properties, but will they be proper arrays? i.e. can I do subObj[currKey].map()?

File: skillmap/src/lib/workspaceProvider.ts [hunk 4]
```diff
@@ -20,10 +23,26 @@ export async function saveProjectAsync(project: pxt.workspace.Project): Promise<
...
+        userState = {
+            id: guidGen(),
+            completedTags: {},
+            mapProgress: {},
+            version: pxt.skillmap.USER_VERSION
+        };
+    }
+    // Read synchronized skillmap state from cloud profile.
+    const skillmapState = await authClient.getSkillmapStateAsync();
+    userState = {
```
Comment (riknoll) [inline skillmap/src/lib/workspaceProvider.ts:40]: this should use the "version" from the authClient

**Review by riknoll — CHANGES_REQUESTED** [review]

File: pxtlib/util.ts (hunk 3 above)
Comment (eanders-MS) [reply pxtlib/util.ts:432]: Zooming out, this feels like something we shouldn't invent. There are edge cases (like creating arrays vs objects, as you point out), and we need unit tests. I generally try to avoid taking additional dependencies, especially client side, but in this case I'd guess that a well-tested version of this algorithm already exists. I'll poke around npm for a good package.

File: skillmap/src/lib/workspaceProvider.ts (hunk 4 above)
Comment (eanders-MS) [reply skillmap/src/lib/workspaceProvider.ts:40]: Can you expand? The meaning of the field hasn't changed, just moved where it is initialized.

File: pxtlib/util.ts (hunk 3 above)
Comment (eanders-MS) [reply pxtlib/util.ts]: Ended up reinventing the wheel anyway. Works with arrays now.

File: pxtlib/auth.ts (hunk 1 above)
Comment (eanders-MS) [reply pxtlib/auth.ts:23]: Reverted this change to point to the staging backend (what it should have been -- prod was incorrect). Identity is working again in staging, and has been updated to support skillmap sync.

File: skillmap/src/lib/workspaceProvider.ts (hunk 4 above)
Comment (eanders-MS) [reply skillmap/src/lib/workspaceProvider.ts:40]: @riknoll My sense is that skillmap should own this version number. Does that square with your thinking?
